python/lsst/multiprofit/modeller.py

```diff
@@ -0,0 +1,840 @@
+            # non-linear IntegralModels
+            if is_psf:
+                n_g = len(gaussians)
+                if n_g != 1:
+                    raise ValueError(f"{component=} has {gaussians=} of len {n_g=}!=1")
+            param_fluxes = component.parameters(paramfilter=g2f.ParamFilter(nonlinear=False, channel=channel))
+            if len(param_fluxes) != 1:
+                raise ValueError(f"Can't make linear source from {component=} with {len(param_fluxes)=}")
+            param_flux: g2f.ParameterD = param_fluxes[0]
+            if param_flux.fixed:
+                gaussians_fixed.append(gaussians.at(0))
+            else:
+                gaussians_free.append((gaussians, param_flux))
+
+        return LinearGaussians(
+            gaussians_fixed=g2.Gaussians(gaussians_fixed), gaussians_free=tuple(gaussians_free)
+        )
+
+
+class FitInputsBase(ABC):
+    """Interface for inputs to a model fit."""
+
+    @abstractmethod
+    def validate_for_model(self, model: Model) -> list[str]:
+        """Check that this FitInputs is valid for a Model.
+
+        Parameters
+        ----------
+        model
+            The model to validate with.
+
+        Returns
+        -------
+        errors
+            A list of validation errors, if any.
+        """
+
+
+class FitInputsDummy(FitInputsBase):
+    """A dummy FitInputs that always fails to validate.
+
+    This class can be used to initialize a FitInputsBase that may be
+    reassigned to a non-dummy derived instance in a loop.
+    """
+
+    def validate_for_model(self, model: Model) -> list[str]:
+        return [
+            "This is a dummy FitInputs and will never validate",
+        ]
+
+
+class FitInputs(FitInputsBase, pydantic.BaseModel):
+    """Model fit inputs for gauss2dfit."""
+
+    model_config = arbitrary_allowed_config
+
+    jacobian: np.ndarray = pydantic.Field(None, title="The full Jacobian array")
+    jacobians: list[list[g2.ImageD]] = pydantic.Field(
+        title="Jacobian arrays (views) for each observation",
+    )
+    outputs_prior: tuple[g2.ImageD, ...] = pydantic.Field(
+        title="Jacobian arrays (views) for each free parameter's prior",
+    )
+    residual: np.ndarray = pydantic.Field(title="The full residual (chi) array")
+    residuals: list[g2.ImageD] = pydantic.Field(
+        default_factory=list,
+        title="Residual (chi) arrays (views) for each observation",
+    )
+    residuals_prior: g2.ImageD = pydantic.Field(
+        title="Shared residual array for all Prior instances",
+    )
+
+    @classmethod
+    def get_sizes(
+        cls,
+        model: Model,
+    ):
+        """Initialize Jacobian and residual arrays for a model.
+
+        Parameters
+        ----------
+        model : `gauss2d.fit.Model`
+            The model to initialize arrays for.
+        """
+        priors = model.priors
+        n_prior_residuals = sum(len(p) for p in priors)
+        params_free = tuple(get_params_uniq(model, fixed=False))
```

Comment:
I added a type hint to `get_params_unique`, although I worry a little about hinting at returns from pybind functions. It may be hard to keep those in sync. mypy would help, of course, but last I checked it's a rather hefty effort to set up mypy stubs for pybind11 modules and without that it just doesn't work.